src/types/common.d.ts

```diff
@@ -109,3 +109,25 @@ export interface ModifierContextState {
   text_line: number;
   top_text_line: number;
 }
+
+
+/* TODO: Move to EasyScore / Grammar / Parser */
+export type Match = string | Match[] | null;
+export interface Rule {
+  // Lexer Rules
+  token?: string; // The token property is a string that is compiled into a RegExp.
+  noSpace?: boolean; // TODO: None of the EasyScore rules specify noSpace, so it is not used anywhere.
+
+  // Parser Rules
+  expect?: RuleFunction[];
+  zeroOrMore?: boolean;
+  oneOrMore?: boolean;
+  maybe?: boolean;
+  or?: boolean;
+  run?: TriggerFunction;
+}
+export interface TriggerState {
+    matches: Match[];
+}
+export type RuleFunction = () => Rule;
+export type TriggerFunction = (state?: TriggerState) => void;
```

Comment:
Actually, all of these belong in `parser.ts` :-)

Reply:
@ronyeh this is my failure I understood from the comment that all of this should be in EasyScore, sorry

Reply:
Yes, I agree that they should be moved back into parser.ts, haha. I think I was working late and a bit sleepy when trying to decide where the interfaces should live.

I think in the end, most of these interfaces should actually be internal to or exported by Parser, as needed. Even the Grammar should actually be an interface (or abstract class) exported by Parser. It might only have one method called `begin()`.... but I think it makes sense that the Parser shouldn't be **importing** any classes **from** EasyScore. EasyScore is just a particular implementation that provides a particular grammar to a generic parser.

Some day, we could conceivably create a EasierScore or a HarderScore class, which provides a different grammar to the same generic parser. Thus, Parser shouldn't be importing anything from the specific implementations.

Once I migrate EasyScore, I'll make sure to remove the `import { Grammar } from './easyscore';` from the top of Parser.ts.

I will make these changes and update this PR.